crates/bevy_ecs/src/component/mod.rs

```diff
@@ -290,9 +289,52 @@ impl Components {
     }
 }
 
-bitflags! {
-    pub struct ComponentFlags: u8 {
-        const ADDED = 1;
-        const MUTATED = 2;
+#[derive(Copy, Clone, Debug)]
+pub struct ComponentCounters {
+    added: u32,
+    changed: u32,
+}
+
+impl ComponentCounters {
+    #[inline]
+    pub fn is_added(&self, system_counter: u32, global_system_counter: u32) -> bool {
+        let component_age = global_system_counter.wrapping_sub(self.added);
+        let system_age = global_system_counter.wrapping_sub(system_counter);
+
+        component_age < system_age
+    }
+
+    #[inline]
+    pub fn is_changed(&self, system_counter: u32, global_system_counter: u32) -> bool {
```

Comment:
it seems like global_system_counter isn't even needed here?
```
global_system_counter - self.changed < global_system_counter - system_counter
// subtract global_system_counter from both sides, which reduces to
- self.changed < - system_counter
// multiply each side by negative one and flip inequality sign
self.changed > system_counter 
```

Reply:
That's what I did at first, but due to overflow handling it limited the detection to only half the range of `u32`.
I think that detecting changes for twice as long is worth adding two subtractions and passing around the global system counter, but that can be discussed.

Reply:
Cool cool. I tried removing it and didn't see any significant perf changes, so this is largely a code-size / complexity tradeoff. I'm not a _huge_ fan of how counters now permeate every api, but I couldn't come up with anything better without making other compromises. I'm cool with leaving this as-is for now, but can we add a comment that calls out why we're doing it this way?

Reply:
Done in 27a7292670292b18052512b4e15e602383e3b261.